File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/method/GetLastOrElse.java
package joookiwi.collection.java.method;

import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.MinimalistCollectionHolder;
import joookiwi.collection.java.annotation.ExtensionFunction;
import joookiwi.collection.java.exception.ImpossibleConstructionException;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;

import java.util.function.IntFunction;
import java.util.function.Supplier;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FAIL_0;

@NotNullByDefault
public final class GetLastOrElse
        extends Utility {

    @Contract(ALWAYS_FAIL_0)
    private GetLastOrElse() { throw new ImpossibleConstructionException("The utility class “GetFirstOrElse” cannot be constructed.", GetLastOrElse.class); }

    //#region -------------------- Facade methods --------------------

    //#region -------------------- predicate (int) → boolean --------------------

    /// Get the **last** element at the specified index in the `collection`
    /// or calling the `defaultValue` function
    /// if it is out of bound of the `collection`
    ///
    /// @param collection   The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param defaultValue The callback to retrieve the default value if it is over the [size][MinimalistCollectionHolder#size] (after calculation)
    @ExtensionFunction
    public static <T extends @Nullable Object> T getLastOrElse(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                               final @NotNull IntFunction<? extends T> defaultValue) {
        if (collection == null)
            return defaultValue.apply(-1);

        final var size = collection.size();
        if (size == 0)
            return defaultValue.apply(-1);
        return collection.get(size - 1);
    }

    /// Get the **last** element at the specified index in the `collection`
    /// or calling the `defaultValue` function
    /// if it is out of bound of the `collection`
    ///
    /// @param collection   The [nullable][Nullable] [collection][CollectionHolder]
    /// @param defaultValue The callback to retrieve the default value if it is over the [size][CollectionHolder#size] (after calculation)
    @ExtensionFunction
    public static <T extends @Nullable Object> T getLastOrElse(final @Nullable CollectionHolder<? extends T> collection,
                                                               final @NotNull IntFunction<? extends T> defaultValue) {
        if (collection == null)
            return defaultValue.apply(-1);
        if (collection.isEmpty())
            return defaultValue.apply(-1);
        return collection.getLast();
    }

    /// Get the **last** element at the specified index in the `collection`
    /// or calling the `defaultValue` function
    /// if it is out of bound of the `collection`
    ///
    /// @param collection   The [nullable][Nullable] collection
    /// @param defaultValue The callback to retrieve the default value if it is over the `size` (after calculation)
    @ExtensionFunction
    public static <T extends @Nullable Object> T getLastOrElse(final T @Nullable [] collection,
                                                               final @NotNull IntFunction<? extends T> defaultValue) {
        if (collection == null)
            return defaultValue.apply(-1);

        final var size = collection.length;
        if (size == 0)
            return defaultValue.apply(-1);
        return collection[size - 1];
    }

    //#endregion -------------------- predicate (int) → boolean --------------------
    //#region -------------------- predicate () → boolean --------------------

    /// Get the **last** element at the specified index in the `collection`
    /// or calling the `defaultValue` function
    /// if it is out of bound of the `collection`
    ///
    /// @param collection   The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param defaultValue The callback to retrieve the default value if it is over the [size][MinimalistCollectionHolder#size] (after calculation)
    @ExtensionFunction
    public static <T extends @Nullable Object> T getLastOrElse(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                               final @NotNull Supplier<? extends T> defaultValue) {
        if (collection == null)
            return defaultValue.get();

        final var size = collection.size();
        if (size == 0)
            return defaultValue.get();
        return collection.get(size - 1);
    }

    /// Get the **last** element at the specified index in the `collection`
    /// or calling the `defaultValue` function
    /// if it is out of bound of the `collection`
    ///
    /// @param collection   The [nullable][Nullable] [collection][CollectionHolder]
    /// @param defaultValue The callback to retrieve the default value if it is over the [size][CollectionHolder#size] (after calculation)
    @ExtensionFunction
    public static <T extends @Nullable Object> T getLastOrElse(final @Nullable CollectionHolder<? extends T> collection,
                                                               final @NotNull Supplier<? extends T> defaultValue) {
        if (collection == null)
            return defaultValue.get();
        if (collection.isEmpty())
            return defaultValue.get();
        return collection.getLast();
    }

    /// Get the **last** element at the specified index in the `collection`
    /// or calling the `defaultValue` function
    /// if it is out of bound of the `collection`
    ///
    /// @param collection   The [nullable][Nullable] collection
    /// @param defaultValue The callback to retrieve the default value if it is over the `size` (after calculation)
    @ExtensionFunction
    public static <T extends @Nullable Object> T getLastOrElse(final T @Nullable [] collection,
                                                               final @NotNull Supplier<? extends T> defaultValue) {
        if (collection == null)
            return defaultValue.get();

        final var size = collection.length;
        if (size == 0)
            return defaultValue.get();
        return collection[size - 1];
    }

    //#endregion -------------------- predicate () → boolean --------------------

    //#endregion -------------------- Facade methods --------------------

}

File: javascript/src/method/getLastOrElse.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Nullable} from "@joookiwi/type"

import type {CollectionHolder}           from "../CollectionHolder"
import type {MinimalistCollectionHolder} from "../MinimalistCollectionHolder"
import type {ReturnCallback}             from "../type/callback"

import {isArray}                       from "./isArray"
import {isArrayByStructure}            from "./isArrayByStructure"
import {isCollectionHolder}            from "./isCollectionHolder"
import {isCollectionHolderByStructure} from "./isCollectionHolderByStructure"
import {isMinimalistCollectionHolder}  from "./isMinimalistCollectionHolder"

//#region -------------------- Facade methods --------------------

/**
 * Get the last element in the {@link collection}
 * or calling the {@link defaultValue} function
 * if it is out of bound of the {@link collection}
 *
 * @param collection   The {@link Nullable nullable} collection ({@link MinimalistCollectionHolder}, {@link CollectionHolder} or {@link ReadonlyArray Array})
 * @param defaultValue The callback to retrieve the default value if it is over the size (after calculation)
 * @extensionFunction
 */
export function getLastOrElse<const T, const U, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>, defaultValue: ReturnCallback<U>,): | T | U
/**
 * Get the last element in the {@link collection}
 * or calling the {@link defaultValue} function
 * if it is out of bound of the {@link collection}
 *
 * @param collection   The {@link Nullable nullable} collection ({@link MinimalistCollectionHolder}, {@link CollectionHolder} or {@link ReadonlyArray Array})
 * @param defaultValue The callback to retrieve the default value if it is over the {@link size} (after calculation)
 * @extensionFunction
 */
export function getLastOrElse<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>, defaultValue: ReturnCallback<T>,): T
export function getLastOrElse<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>, defaultValue: ReturnCallback<unknown>,) {
    if (collection == null)
        return defaultValue()
    if (isCollectionHolder(collection,))
        return getLastOrElseByCollectionHolder(collection, defaultValue,)
    if (isArray(collection,))
        return getLastOrElseByArray(collection, defaultValue,)
    if (isMinimalistCollectionHolder(collection,))
        return getLastOrElseByMinimalistCollectionHolder(collection, defaultValue,)

    if (isCollectionHolderByStructure<T>(collection,))
        return getLastOrElseByCollectionHolder(collection, defaultValue,)
    if (isArrayByStructure<T>(collection,))
        return getLastOrElseByArray(collection, defaultValue,)
    return getLastOrElseByMinimalistCollectionHolder(collection, defaultValue,)
}


/**
 * Get the last element in the {@link collection}
 * or calling the {@link defaultValue} function
 * if it is out of bound of the {@link collection}
 *
 * @param collection   The nullable collection
 * @param defaultValue The callback to retrieve the default value if it is over the {@link size} (after calculation)
 * @extensionFunction
 */
export function getLastOrElseByMinimalistCollectionHolder<const T, const U, >(collection: Nullable<MinimalistCollectionHolder<T>>, defaultValue: ReturnCallback<U>,): | T | U
/**
 * Get the last element in the {@link collection}
 * or calling the {@link defaultValue} function
 * if it is out of bound of the {@link collection}
 *
 * @param collection   The nullable collection
 * @param defaultValue The callback to retrieve the default value if it is over the {@link size} (after calculation)
 * @extensionFunction
 */
export function getLastOrElseByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>, defaultValue: ReturnCallback<T>,): T
export function getLastOrElseByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>, defaultValue: ReturnCallback<unknown>,) {
    if (collection == null)
        return defaultValue()

    const size = collection.size
    if (size == 0)
        return defaultValue()
    return collection.get(collection.size - 1,)
}

/**
 * Get the last element in the {@link collection}
 * or calling the {@link defaultValue} function
 * if it is out of bound of the {@link collection}
 *
 * @param collection   The nullable collection
 * @param defaultValue The callback to retrieve the default value if it is over the {@link size} (after calculation)
 * @extensionFunction
 */
export function getLastOrElseByCollectionHolder<const T, const U, >(collection: Nullable<CollectionHolder<T>>, defaultValue: ReturnCallback<U>,): | T | U
/**
 * Get the last element in the {@link collection}
 * or calling the {@link defaultValue} function
 * if it is out of bound of the {@link collection}
 *
 * @param collection   The nullable collection
 * @param defaultValue The callback to retrieve the default value if it is over the {@link size} (after calculation)
 * @extensionFunction
 */
export function getLastOrElseByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>, defaultValue: ReturnCallback<T>,): T
export function getLastOrElseByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>, defaultValue: ReturnCallback<unknown>,) {
    if (collection == null)
        return defaultValue()
    if (collection.isEmpty)
        return defaultValue()
    return collection.getLast()
}


/**
 * Get the last element in the {@link collection}
 * or calling the {@link defaultValue} function
 * if it is out of bound of the {@link collection}
 *
 * @param collection   The nullable collection
 * @param defaultValue The callback to retrieve the default value if it is over the {@link size} (after calculation)
 * @extensionFunction
 */
export function getLastOrElseByArray<const T, const U, >(collection: Nullable<readonly T[]>, defaultValue: ReturnCallback<U>,): | T | U
/**
 * Get the last element in the {@link collection}
 * or calling the {@link defaultValue} function
 * if it is out of bound of the {@link collection}
 *
 * @param collection   The nullable collection
 * @param defaultValue The callback to retrieve the default value if it is over the {@link size} (after calculation)
 * @extensionFunction
 */
export function getLastOrElseByArray<const T, >(collection: Nullable<readonly T[]>, defaultValue: ReturnCallback<T>,): T
export function getLastOrElseByArray<const T, >(collection: Nullable<readonly T[]>, defaultValue: ReturnCallback<unknown>,) {
    if (collection == null)
        return defaultValue()
    const size = collection.length
    if (size == 0)
        return defaultValue()
    return collection[collection.length - 1]
}

//#endregion -------------------- Facade methods --------------------
